src/Juvix/Core/IR/Typechecker.hs

```diff
@@ -23,67 +74,216 @@ typeTerm ∷
+      if (i >= j)
+        then
+          ( do
+              logOutput $
+                failed
+                  <> "The input annotation is of * level"
+                  <> show j
+                  <> ", which is not greater than the term's * level "
+                  <> show i
+                  <> ". "
+              throw @"typecheckError" (UniverseMismatch t (snd ann))
+          )
+        else
+          ( do
+              logOutput $
+                passed
+                  <> "The input annotation is of * level"
+                  <> show j
+                  <> ", which is greater than the term's * level "
+                  <> show i
+                  <> ". "
+                  <> typechecked t ann
+              return ()
+          )
+    _ → do
+      logOutput $
+        failed
+          <> " The input annotation is not of * type, it is of type "
+          <> show (snd ann)
+      throw @"typecheckError" (ShouldBeStar (snd ann))
+  -- *-Pi (Universe introduction rule)
+typeTerm p ii g t@(Pi _pi varType resultType) ann = do
+  typeTermIntroLog t ann
+  logOutput "patterned matched to be a dependent function type term. Type checker applies the universe introduction rule. Checking that sigma is zero."
+  unless -- checks sigma = 0.
+    (SNat 0 == fst ann)
+    ( do
+        logOutput $
+          failed
+            <> "The input usage (sigma) is "
+            <> show (fst ann)
+            <> ", which is not zero. "
+        throw @"typecheckError" SigmaMustBeZero
+    )
+  logOutput $
+    passed
+      <> "The input usage (sigma) is "
+      <> show (fst ann)
+      <> ", it is zero, as required. "
+  logOutput $
+    "Checking that the input type "
+      <> show (snd ann)
+      <> "is *i. "
   case snd ann of
     VStar _ → do
-      typeTerm p ii g varType ann -- checks varType is of type Star i
-      ty ← evalTerm p varType []
+      logOutput $
+        passed
+          <> "The input type is "
+          <> show (snd ann)
+          <> ", it is *i, as required. "
+      logOutput $
+        "Checking that the variable (V) type checked to the input annotation, which is just type checked to be of usage 0 and type *i. "
+      typeTerm p ii g varType ann
+      logOutput $
+        passed
+          <> "The variable (V) is "
+          <> show (fst ann)
+          <> " and type "
+          <> show (snd ann)
+          <> ", as required. Checking that the result (R), with x of type V in the context, type checked against the input annotation. "
+      ty ← evalTerm p varType [] -- V
       typeTerm
-        p -- checks resultType is of type Star i
+        p -- param
         (ii + 1)
-        ((Local ii, (pi, ty)) : g)
-        (substTerm 0 (Free (Local ii)) resultType)
-        ann
-    _ →
-      throw @"typecheckError" UniverseLevelMustMatch
+        ((Local ii, (SNat 0, ty)) : g) -- add x of type V, with zero usage to the context
```

Comment:
According to p.22 of the lang ref, x should be added to the context with 0 usage.